### DISCUSSION TITLE
how to Deserialize BSON Object

### DISCUSSION THREAD

**Discussion by mirkojpn87** [nlohmann/json#2884] [Q&A] [answered]
hi.
i have a couple Bson serialized Databases with C# and MongoDB library.
under C# i'm able to deserialize it like this : 

```
static void Main(string[] args)
    {
        foreach (string file in Directory.EnumerateFiles(Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location), "*.bson"))
        {
            byte[] contents =
                File.ReadAllBytes(file);

            Databases cl = BsonSerializer.Deserialize<Databases>(contents);
            Console.WriteLine(cl._databases["001"]._date.ToString());
        }
    }

    private class Databases
    {
        public Dictionary<string, SampleClass> _databases;
    }

```

i'm looking to do the same in C++ is possible?

**Comment by nlohmann** [comment, marked as answer]
Yes, this should be possible. For the conversion of user-defined types (like `Databases` in your case), see https://json.nlohmann.me/features/arbitrary_types/. Anything related to BSON is documented here: https://json.nlohmann.me/features/binary_formats/bson/

**Reply by mirkojpn87** [reply]
hi nlhmann, thank you for your help.

so, from what i understand i should do my own implementation of "from_bson()"?

**Reply by nlohmann** [reply]
No, the translation into/from BSON is always the same. You can defined a user-defined mapping to/from JSON though.

**Reply by mirkojpn87** [reply]
Hi.
i'm working on it...but i have some troubles to figure out how to make it works.

what i have is this 

![Screenshot 2021-07-26 123607](https://user-images.githubusercontent.com/549026/126950735-de7dbd97-5886-4dbb-b131-1e63de1cf97e.png)
![Screenshot 2021-07-26 123615](https://user-images.githubusercontent.com/549026/126950747-b17469b8-12b2-4c41-a51e-a40afd0c81f4.png)


what i'm missing out?

**Reply by gregmarr** [reply]
It doesn't look like you've implemented `to_json` for `EventCondition`.

**Reply by mirkojpn87** [reply]
since i'm not going to use everytime the same class ( that in this case is EventCondition) but a lot of other, i have to implement `to_json` for every class that i'm going to use?

**Reply by nlohmann** [reply]
Yes

**Reply by gregmarr** [reply]
Until C++ gets static reflection and code generation in C++23 or later, there is no way for the library to automatically (de-)serialize a class that doesn't have that function.
